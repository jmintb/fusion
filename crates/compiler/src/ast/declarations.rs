use super::{
    identifiers::{DeclarationID, FunctionDeclarationID, StructDeclarationID},
    nodes::Identifier,
};

#[derive(Default, Clone, Debug)]
pub struct ModuleDeclaration {
    pub struct_declarations: Vec<StructDeclarationID>,
    pub function_declarations: Vec<FunctionDeclarationID>,
    pub identifier: Identifier,
}

impl ModuleDeclaration {
    pub fn new(identifier: Identifier) -> Self {
        Self {
            struct_declarations: Vec::new(),
            function_declarations: Vec::new(),
            identifier,
        }
    }

    pub fn declare(&mut self, id: DeclarationID) {
        // TODO: collisions?
        match id {
            DeclarationID::StructDeclaration(id) => {
                self.struct_declarations.push(id);
            }
            DeclarationID::FunctionDeclaration(id) => {
                self.function_declarations.push(id);
            }
            DeclarationID::ModuleDeclarationID(_id) => todo!(),
        }
    }
}
