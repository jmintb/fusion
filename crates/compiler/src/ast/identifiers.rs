use crate::identifiers::ID;

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash, Debug)]
pub struct ModuleDeclarationID(pub ID);

impl From<ModuleDeclarationID> for NodeID {
    fn from(value: ModuleDeclarationID) -> Self {
        Self::Declaration(DeclarationID::ModuleDeclarationID(value))
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash, Debug)]
pub struct StructDeclarationID(pub ID);

impl From<StructDeclarationID> for NodeID {
    fn from(value: StructDeclarationID) -> Self {
        Self::Declaration(DeclarationID::StructDeclaration(value))
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash, Debug, Default)]
pub struct FunctionDeclarationID(pub ID);

impl From<ID> for FunctionDeclarationID {
    fn from(value: ID) -> Self {
        Self(value)
    }
}

impl From<FunctionDeclarationID> for NodeID {
    fn from(value: FunctionDeclarationID) -> Self {
        Self::Declaration(DeclarationID::FunctionDeclaration(value))
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash, Debug)]
pub struct ExpressionID(pub ID);

impl From<ID> for ExpressionID {
    fn from(value: ID) -> Self {
        Self(value)
    }
}

impl From<ExpressionID> for NodeID {
    fn from(value: ExpressionID) -> Self {
        Self::Statement(StatementID::Expression(value))
    }
}

impl From<ExpressionID> for StatementID {
    fn from(value: ExpressionID) -> Self {
        Self::Expression(value)
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Debug, Hash)]
pub struct BlockID(pub ID);

impl From<usize> for BlockID {
    fn from(value: usize) -> Self {
        Self(value)
    }
}

impl From<BlockID> for NodeID {
    fn from(value: BlockID) -> Self {
        Self::Block(value)
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Debug, Hash)]
pub enum StatementID {
    Declaration(DeclarationID),
    Expression(ExpressionID),
}

impl From<FunctionDeclarationID> for StatementID {
    fn from(value: FunctionDeclarationID) -> Self {
        StatementID::Declaration(DeclarationID::FunctionDeclaration(value))
    }
}

impl From<StructDeclarationID> for StatementID {
    fn from(value: StructDeclarationID) -> Self {
        StatementID::Declaration(DeclarationID::StructDeclaration(value))
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash, Debug)]
pub enum NodeID {
    Declaration(DeclarationID),
    Block(BlockID),
    Statement(StatementID),
    Scope(ScopeID),
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash, Debug)]
pub enum DeclarationID {
    StructDeclaration(StructDeclarationID),
    FunctionDeclaration(FunctionDeclarationID),
    ModuleDeclarationID(ModuleDeclarationID),
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
pub struct TypeID(usize);

impl From<usize> for TypeID {
    fn from(value: usize) -> Self {
        Self(value)
    }
}

impl From<StructDeclarationID> for usize {
    fn from(val: StructDeclarationID) -> Self {
        val.0
    }
}

impl From<ModuleDeclarationID> for usize {
    fn from(val: ModuleDeclarationID) -> Self {
        val.0
    }
}

impl From<usize> for StructDeclarationID {
    fn from(value: usize) -> Self {
        Self(value)
    }
}

impl From<usize> for ModuleDeclarationID {
    fn from(value: usize) -> Self {
        Self(value)
    }
}

impl From<TypeID> for usize {
    fn from(val: TypeID) -> Self {
        val.0
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
pub struct FunctionID(usize);
impl From<usize> for FunctionID {
    fn from(value: usize) -> Self {
        Self(value)
    }
}

impl From<FunctionID> for usize {
    fn from(val: FunctionID) -> Self {
        val.0
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
pub struct VariableID(usize);
impl From<usize> for VariableID {
    fn from(value: usize) -> Self {
        Self(value)
    }
}

impl From<VariableID> for usize {
    fn from(val: VariableID) -> Self {
        val.0
    }
}
#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Debug, Hash)]
pub struct ScopeID(pub usize);

impl From<ID> for ScopeID {
    fn from(value: ID) -> Self {
        Self(value)
    }
}

impl From<ScopeID> for NodeID {
    fn from(value: ScopeID) -> Self {
        Self::Scope(value)
    }
}
