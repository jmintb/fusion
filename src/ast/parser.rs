use anyhow::{bail, Result};
use pest::{
    iterators::{Pair, Pairs},
    Parser,
};
use pest_derive::Parser;

use crate::ast::{
    identifiers::DeclarationID,
    nodes::{AccessModes, FunctionKeyword, Identifier},
};

use super::{
    declarations::ModuleDeclaration,
    identifiers::{ExpressionID, FunctionDeclarationID, StatementID, StructDeclarationID},
    nodes::{
        Assignment, Block, Call, Expression, FunctionArg, FunctionDeclaration, Integer, Return,
        StructDeclaration, StructField, StructInit, Type, Value, Operation, Operator
    },
    Ast, NodeDatabase,
};
use tracing::debug;

#[derive(Parser)]
#[grammar = "typed_script.pest"]
struct TSParser;

struct AstBuilder {
    db: NodeDatabase,
    ast: Ast,
}

impl AstBuilder {
    fn new() -> Self {
        Self {
            db: NodeDatabase::default(),
            ast: Ast::new()
        }
    }
}

pub fn parse(input: &str) -> Result<(Ast, NodeDatabase, Vec<String>)> {
    let program = TSParser::parse(Rule::program, &input)?;
    let mut builder = AstBuilder::new();
    parse_program(program, &mut builder)
}

fn parse_program(
    program: Pairs<'_, Rule>,
    mut builder: &mut AstBuilder,
) -> Result<(Ast, NodeDatabase, Vec<String>)> {
    let mut module = ModuleDeclaration::new(Identifier::new("main".to_string()));

    for rule in program.into_iter() {
        let id = match rule.as_rule() {
            Rule::program => todo!(),
            Rule::function => {
                let id = parse_function_decl(builder, rule)?;
                DeclarationID::FunctionDeclaration(id)
            }
            Rule::r#struct => {
                let id = parse_struct_decl(builder, rule)?;
                DeclarationID::StructDeclaration(id)
            }
            Rule::EOI => break,
            _ => panic!("unexpected rule {:?}", rule.as_rule()),
        };
        module.declare(id);
    }

    let module_id = builder.db.new_module_declaration(module);
    builder.ast.modules.push(module_id);

    Ok((builder.ast.clone(), builder.db.clone(), Vec::new()))
}

fn parse_struct_decl(
    builder: &mut AstBuilder,
    rule: pest::iterators::Pair<'_, Rule>,
) -> Result<StructDeclarationID> {
    let mut decl = rule.into_inner();

    let identifier = Identifier::new(decl.next().unwrap().as_str().to_string());

    let fields = decl
        .map(parse_struct_field_declaration)
        .collect::<Result<Vec<StructField>>>()?;

    Ok(builder
        .db
        .new_struct_declaration(StructDeclaration { identifier, fields }))
}

fn parse_struct_field_declaration(decl: Pair<Rule>) -> Result<StructField> {
    let mut decl = decl.into_inner();

    let identifier = Identifier::new(decl.next().unwrap().as_str().to_string());

    let r#type = parse_type(decl.next().unwrap())?;

    Ok(StructField { identifier, r#type })
}

fn parse_function_decl(
    builder: &mut AstBuilder,
    rule: Pair<Rule>,
) -> Result<FunctionDeclarationID> {
    let mut decl = rule.into_inner();
    let next = decl.next().unwrap();

    let keywords = if let Rule::functionType = next.as_rule() {
        vec![FunctionKeyword::try_from(next.as_str().to_string())?]
    } else {
        Vec::new()
    };

    let identifier = Identifier::new(decl.next().unwrap().as_str().to_string());

    let mut next = decl.next();

    let function_args: Vec<FunctionArg> =
        if let Some(Rule::functionArgs) = next.clone().map(|next| next.as_rule()) {
            let args = next.clone().map(|next| next.into_inner()).unwrap();
            next = decl.next();
            args.map(parse_fn_arg)
                .collect::<Result<Vec<FunctionArg>>>()?
        } else {
            vec![]
        };

    let return_type = if let Some(Rule::r#type) = next.clone().map(|next| next.as_rule()) {
        let ty = parse_type(next.unwrap())?;
        next = decl.next();
        Some(ty)
    } else {
        None
    };

    let body = if let Some(Rule::functionBody) = next.clone().map(|next| next.as_rule()) {
        let body = next
            .map(|next| next.into_inner())
            .unwrap()
            .map(|statement| parse_statement(builder, statement))
            .collect::<Result<Vec<StatementID>>>()?;
        Some(builder.db.new_block(Block::new(body)))
    } else {
        None
    };

    Ok(builder.db.new_function_declaration(FunctionDeclaration {
        keywords,
        identifier,
        arguments: function_args,
        body,
        return_type,
    }))
}

fn parse_statement(
    builder: &mut AstBuilder,
    statement: Pair<Rule>,
) -> Result<StatementID, anyhow::Error> {
    Ok(match statement.as_rule() {
        Rule::function => parse_function_decl(builder, statement)?.into(),
        Rule::r#struct => parse_struct_decl(builder, statement)?.into(),
        Rule::call
        | Rule::structInit
        | Rule::string
        | Rule::if_else
        | Rule::while_loop
        | Rule::r#return
        | Rule::assignment
        | Rule::assign => parse_expression(builder, statement)?.into(),
        _ => bail!("Recieved unexpected rule: {:?}", statement),
    })
}

fn parse_expression(builder: &mut AstBuilder, pair: Pair<Rule>) -> Result<ExpressionID> {
    let expression = match pair.as_rule() {
        Rule::string => Expression::Value(parse_string(pair)?),
        Rule::assignment => Expression::Assignment(parse_assignment(builder, pair)?),
        Rule::identifier => Expression::Value(Value::Variable(Identifier::new(
            pair.as_str().trim().into(),
        ))),
        Rule::call => Expression::Call(parse_fn_call(builder, pair)?),
        Rule::structInit => Expression::Struct(parse_struct_init(builder, pair)?),
        // Rule::structFieldRef => parse_struct_field_ref(expression)?,
        Rule::integer => Expression::Value(Value::Integer(parse_integer(pair)?)),
        // Rule::reference => {
        //     Expression::Value(parse_string(expression.into_inner().next().unwrap())?)
        // }
        Rule::operation => Expression::Operation(parse_operation(builder, pair)?),
        // Rule::boolean => Expression::Value(parse_boolean(expression)?),
        // Rule::r#if_else => Expression::If(parse_if(expression)?),
        // Rule::while_loop => Expression::While(parse_while(expression)?),
        // Rule::assign => Expression::Assign(parse_assign(expression)?),
        Rule::r#return => Expression::Return(parse_return(builder, pair)?),
        // Rule::array => Expression::Array(parse_array(expression)?),
        // Rule::array_lookup => Expression::ArrayLookup(parse_array_lookup(expression)?),
        _ => panic!("Got unexpected expression: {:?}", pair.as_rule()),
    };

    let expression_id = builder.db.new_expression(expression);

    Ok(expression_id)
}

fn parse_operation(builder: &mut AstBuilder, pair: Pair<Rule>) -> Result<Operation> {
    let mut inner = if let Rule::operation = pair.as_rule() {
        pair.into_inner()
    } else {
        bail!("expected operation rule got {:?}", pair);
    };

    let first_operand = parse_expression(builder, inner.next().unwrap())?;

    let operator = parse_operator(inner.next().unwrap())?;

    let second_operand = parse_expression(builder, inner.next().unwrap())?;

    Ok(Operation::Binary(
        first_operand,
        operator,
        second_operand,
    ))
}

fn parse_operator(expression: Pair<Rule>) -> Result<Operator> {
    Ok(match expression.as_rule() {
        Rule::addition => Operator::Addition,
        Rule::subtraction => Operator::Subtraction,
        Rule::division => Operator::Division,
        Rule::multiplication => Operator::Multiplication,
        Rule::equal => Operator::Equality,
        Rule::not_equal => Operator::Inequality,
        Rule::greater_than => Operator::GreaterThan,
        Rule::less_than => Operator::LessThan,
        Rule::greater_than_or_equal => Operator::GreaterThanOrEqual,
        Rule::less_than_or_equal => Operator::LessThanOrEqual,
        _ => bail!("expected an infix operator but got: {:?}", expression),
    })
}

fn parse_return(builder: &mut AstBuilder, pair: Pair<Rule>) -> Result<Return> {
    let mut inner = if let Rule::r#return = pair.as_rule() {
        pair.into_inner()
    } else {
        bail!("expected return rule found {}", pair.as_str());
    };

    let expression_id = if let Some(next) = inner.next() {
        Some(parse_expression(builder, next)?.into())
    } else {
        None
    };

    Ok(Return {
        expression: expression_id,
    })
}

fn parse_struct_init(builder: &mut AstBuilder, pair: Pair<Rule>) -> Result<StructInit> {
    todo!()
}

fn parse_integer(integer: Pair<Rule>) -> Result<Integer> {
    if let Rule::integer = integer.as_rule() {
        let value = integer.as_str().trim().parse()?;
        Ok(Integer {
            value,
            signed: true,
            size: 32,
        })
    } else {
        bail!("expected string , got {:?}", integer.as_rule())
    }
}

fn parse_fn_call(builder: &mut AstBuilder, call_expression: Pair<Rule>) -> Result<Call> {
    let mut inner = call_expression.clone().into_inner().into_iter();

    let id = inner.next().unwrap();

    let args = inner.next().unwrap().into_inner().into_iter();

    let args = args
        .map(|arg| parse_expression(builder, arg).unwrap())
        .collect();


    let call = Call { function_id: Identifier(id.as_str().trim().to_string()),
        arguments: args,
    };

    Ok(call)
}

fn parse_assignment(builder: &mut AstBuilder, pair: Pair<'_, Rule>) -> Result<Assignment> {
    let mut inner_rules = pair.into_inner();

    let identifier = inner_rules.next().unwrap();

    let expression = inner_rules.next().unwrap();

    let expression = parse_expression(builder, expression);

    let assignment = Assignment {
        id: Identifier::new(identifier.as_str().trim().into()),
        expression: expression.unwrap(),
    };

    Ok(assignment)
}

fn parse_string(string: Pair<Rule>) -> Result<Value> {
    if let Rule::string = string.as_rule() {
        Ok(Value::String(string.into_inner().as_str().to_string()))
    } else {
        bail!("expected string , got {:?}", string.as_rule())
    }
}

fn parse_fn_arg(arg: Pair<Rule>) -> Result<FunctionArg> {
    let mut inner_rules = arg.into_inner();
    let n_rules = inner_rules.len();
    let Some(first_rule) = inner_rules.next() else {
        bail!("function argument was empty");
    };

    let access_mode = if let Rule::accessMode = first_rule.as_rule() {
        match first_rule
            .clone()
            .into_inner()
            .next()
            .map(|inner| inner.as_rule())
        {
            Some(Rule::letAccess) => AccessModes::Let,
            Some(Rule::ownedAccess) => AccessModes::Owned,
            Some(Rule::inoutAccess) => AccessModes::Inout,
            e @ _ => bail!("exptec to find an access found instead foud {:?}", e),
        }
    } else {
        AccessModes::default()
    };

    let name = if n_rules == 2 {
        Identifier::new(first_rule.as_str().to_string())
    } else {
        Identifier::new(inner_rules.next().unwrap().as_str().to_string())
    };

    Ok(FunctionArg {
        access_mode,
        name,
        r#type: inner_rules.next().map(|ty_id| parse_type(ty_id).unwrap()),
    })
}

fn parse_type(ty: Pair<Rule>) -> Result<Type> {
    let mut inner = ty.into_inner();

    let next = inner.next().unwrap();
    Ok(match next.as_rule() {
        Rule::signed_integer => Type::SignedInteger,
        Rule::unsigned_integer => Type::UnsignedInteger,
        Rule::array_type => {
            let mut next_inner = next.into_inner();
            let item_type = parse_type(next_inner.next().unwrap())?;
            let len = next_inner.next().unwrap().as_str().parse::<usize>()?;
            Type::Array(item_type.into(), len)
        }
        Rule::string_type => Type::StringLiteral,
        Rule::named_type => Type::Named(Identifier::new(
            next.into_inner().next().unwrap().as_str().to_string(),
        )),
        Rule::pointer => Type::Pointer,
        _ => bail!("expected a type rule got: {:?}", next),
    })
}

#[cfg(test)]
mod test {
    use super::*;
    use anyhow::Result;
    use rstest::rstest;
    use std::path::PathBuf;
    use tracing::debug;

    #[rstest]
    #[test_log::test]
    fn snapshop_ast_output(#[files("./test_programs/test_*.ts")] path: PathBuf) -> Result<()> {
        let input = std::fs::read_to_string(&path)?;
        let ast = parse(&input)?;
        debug!("input for parser:\n {input}");
        insta::assert_snapshot!(
            format!(
                "test_ast_snapshot_{}",
                path.file_name().unwrap().to_str().unwrap()
            ),
            format!("{}", ast.0.to_string(&ast.1))
        );
        Ok(())
    }
}
