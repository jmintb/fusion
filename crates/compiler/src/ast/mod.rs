pub mod declarations;
pub mod identifiers;
pub mod nodes;
pub mod parser;
pub mod scopes;

use anyhow::{bail, Result};
use std::collections::{HashMap, VecDeque};
use tracing::debug;

use crate::identifiers::{IDGenerator, ID};

use self::{
    declarations::ModuleDeclaration,
    identifiers::{BlockID, ExpressionID, ScopeID, StatementID},
    nodes::{
        Block, Expression, FunctionDeclaration, Identifier, IfElseStatement, IfStatement,
        StructDeclaration, While,
    },
};
use super::ast::identifiers::{
    DeclarationID, FunctionDeclarationID, ModuleDeclarationID, NodeID, StructDeclarationID,
};

type WalkerFn<Ctx> = dyn FnMut(&NodeDatabase, NodeID, Option<NodeID>, &mut Ctx) -> Result<()>;

#[derive(Debug, Clone)]
pub struct Scope {
    pub nodes: Vec<NodeID>,
    pub parent_scope: Option<ScopeID>,
    pub associated_node: Option<NodeID>,
}

#[derive(Debug, Clone)]
pub struct Ast {
    modules: Vec<ModuleDeclarationID>,
}

impl Default for Ast {
    fn default() -> Self {
        Self::new()
    }
}

impl Ast {
    pub fn new() -> Self {
        Self {
            modules: Vec::new(),
        }
    }

    pub fn get_entry_function_id(&self, db: &NodeDatabase) -> Result<FunctionDeclarationID> {
        db.get_function_declaration_id_from_identifier(Identifier("main".to_string()))
    }

    pub fn get_node_relationships(&self, db: &NodeDatabase) -> HashMap<NodeID, Vec<NodeID>> {
        let mut output: (HashMap<NodeID, Vec<NodeID>>, Option<NodeID>) = (HashMap::new(), None);
        let mut walker =
            |_db: &NodeDatabase,
             node_id: NodeID,
             parent_node_id: Option<NodeID>,
             output: &mut (HashMap<NodeID, Vec<NodeID>>, Option<NodeID>)| {
                if let Some(parent_id) = parent_node_id {
                    output
                        .0
                        .entry(parent_id)
                        .and_modify(|children| children.push(node_id))
                        .or_insert(vec![node_id]);
                }

                output.1 = parent_node_id;

                Ok(())
            };

        self.traverse(db, &mut walker, &mut output);

        output.0
    }

    // TODO: this kind of has too much detail now. Ideally we only want the structure of the tree.
    pub fn to_string(&self, db: &NodeDatabase) -> String {
        let mut output: (HashMap<NodeID, Vec<NodeID>>, Option<NodeID>) = (HashMap::new(), None);
        let mut walker =
            |_db: &NodeDatabase,
             node_id: NodeID,
             parent_node_id: Option<NodeID>,
             output: &mut (HashMap<NodeID, Vec<NodeID>>, Option<NodeID>)| {
                if let Some(parent_id) = parent_node_id {
                    output
                        .0
                        .entry(parent_id)
                        .and_modify(|children| children.push(node_id))
                        .or_insert(vec![node_id]);
                }

                output.1 = parent_node_id;

                Ok(())
            };

        self.traverse(db, &mut walker, &mut output);

        let mut tree_rows = vec!["root".to_string()];

        let mut que: VecDeque<Vec<Vec<NodeID>>> = vec![self
            .modules
            .clone()
            .into_iter()
            .map(|module_id| vec![NodeID::from(module_id)])
            .collect()]
        .into();

        while let Some(parent_row) = que.pop_front() {
            let parent_names: Vec<Vec<String>> = parent_row
                .iter()
                .filter(|parent| !parent.is_empty())
                .map(|parent| {
                    parent
                        .iter()
                        .rev()
                        .map(|parent| match parent {
                            &NodeID::Declaration(DeclarationID::ModuleDeclarationID(id)) => {
                                format!("module_{}", id.0)
                            }
                            &NodeID::Block(BlockID(id)) => format!("block_{id}"),
                            &NodeID::Statement(StatementID::Expression(id)) => db
                                .expressions
                                .get(&id)
                                .unwrap()
                                .to_debug_string(db)
                                .unwrap(),
                            &NodeID::Statement(StatementID::Declaration(_)) => db
                                .ids
                                .get(parent)
                                .map(|id| id.0.clone())
                                .unwrap_or("unnamed".to_string())
                                .clone(),
                            &NodeID::Declaration(_) => db
                                .ids
                                .get(parent)
                                .map(|id| id.0.clone())
                                .unwrap_or("unnamed".to_string())
                                .clone(),
                            _ => todo!("format!({parent:?})"),
                        })
                        .collect::<Vec<String>>()
                })
                .collect();

            debug!("parent names {:?}", parent_names);
            if !parent_names.is_empty() {
                tree_rows.push(format!("{:#?}", parent_names));
            }

            let next_row: Vec<Vec<NodeID>> = parent_row
                .iter()
                .flatten()
                .map(|parent| output.0.get(parent).unwrap_or(&Vec::new()).clone())
                .collect();

            if next_row.is_empty() {
                break;
            }

            que.push_back(next_row);
        }

        tree_rows.join(", ").to_string()
    }

    pub fn traverse<Ctx>(
        &self,
        db: &NodeDatabase,
        walker: &mut WalkerFn<Ctx>,
        walker_context: &mut Ctx,
    ) {
        let mut queue: VecDeque<(NodeID, Option<NodeID>)> = self
            .modules
            .iter()
            .map(|module_id| {
                (
                    NodeID::Declaration(DeclarationID::ModuleDeclarationID(*module_id)),
                    None,
                )
            })
            .collect();

        let process_block =
            |block_id: &BlockID,
             parent_id: Option<NodeID>,
             queue: &mut VecDeque<(NodeID, Option<NodeID>)>| {
                let block = db.blocks.get(block_id).unwrap();
                for &statement in block.statements.iter() {
                    match statement {
                        StatementID::Declaration(_decl) => todo!(),
                        StatementID::Expression(expression_id) => {
                            queue.push_front((expression_id.into(), parent_id));
                        }
                    }
                }
            };

        // TODO: this is not parralellised
        while let Some((next, parent)) = queue.pop_front() {
            match next {
                NodeID::Declaration(DeclarationID::ModuleDeclarationID(module_id)) => {
                    walker(db, module_id.into(), parent, walker_context).unwrap();
                    let module_declaration = db.module_declarations.get(&module_id).unwrap();
                    for &struct_declaration in module_declaration.struct_declarations.iter() {
                        queue.push_front((struct_declaration.into(), Some(module_id.into())));
                    }

                    for &function_declaration in module_declaration.function_declarations.iter() {
                        queue.push_front((function_declaration.into(), Some(module_id.into())));
                    }
                }

                NodeID::Declaration(DeclarationID::FunctionDeclaration(function_id)) => {
                    walker(db, function_id.into(), parent, walker_context).unwrap();
                    let function_declaration = db.function_declarations.get(&function_id).unwrap();

                    if let Some(body) = function_declaration.body {
                        queue.push_front((body.into(), Some(function_id.into())));
                    }
                }
                NodeID::Block(block_id) => {
                    walker(db, block_id.into(), parent, walker_context).unwrap();
                    process_block(&block_id, Some(block_id.into()), &mut queue);
                }

                // TODO: All expressions need to be "unpacked" and processed as they can contain blocks.

                // NEXT: Get If Ellse statement snapshots to include if/else blocks
                NodeID::Statement(StatementID::Expression(expression_id)) => {
                    let expression = db.expressions.get(&expression_id).unwrap();

                    match expression {
                        Expression::If(IfStatement {
                            condition: _, // TODO:conditions need  to be pressed by walker.
                            then_block,
                        }) => process_block(then_block, parent, &mut queue),
                        Expression::Ifelse(IfElseStatement {
                            condition,
                            then_block,
                            else_block,
                        }) => {
                            debug!("process ifelse statement, with parent {:?}", parent);
                            walker(db, expression_id.into(), parent, walker_context).unwrap();
                            queue.push_front(((*condition).into(), parent));
                            queue.push_front(((*then_block).into(), Some(expression_id.into())));
                            queue.push_front(((*else_block).into(), Some(expression_id.into())));
                        }
                        Expression::While(While { condition, body }) => {
                            queue.push_front(((*condition).into(), parent));
                            process_block(body, parent, &mut queue);
                        }
                        _expression => {
                            walker(db, expression_id.into(), parent, walker_context).unwrap();
                        }
                    }
                }
                decl => todo!("{decl:?}"),
            }
        }
    }
}

#[derive(Default, Clone, Debug)]
pub struct NodeDatabase {
    pub module_declarations: HashMap<ModuleDeclarationID, ModuleDeclaration>,
    pub struct_declarations: HashMap<StructDeclarationID, StructDeclaration>,
    pub function_declarations: HashMap<FunctionDeclarationID, FunctionDeclaration>,
    pub expressions: HashMap<ExpressionID, Expression>,
    pub blocks: HashMap<BlockID, Block>,
    pub ids: HashMap<NodeID, Identifier>,
    pub id_generator: IDGenerator,
}

impl NodeDatabase {
    pub fn get_function_declaration_id_from_identifier(
        &self,
        name: impl Into<Identifier>,
    ) -> Result<FunctionDeclarationID> {
        let name: Identifier = name.into();
        for (function_declaration_id, function_declaration) in self.function_declarations.iter() {
            if function_declaration.identifier == name {
                return Ok(*function_declaration_id);
            }
        }

        bail!(format!(
            "failed to find function declaration ID for identifier {name:?}"
        ));
    }

    fn new_module_declaration(&mut self, declaration: ModuleDeclaration) -> ModuleDeclarationID {
        let id = self.new_node_id_for_identifier(declaration.identifier.clone());
        self.module_declarations.insert(id, declaration);

        id
    }

    fn new_struct_declaration(&mut self, declaration: StructDeclaration) -> StructDeclarationID {
        // TODO: what to do about name collisions?
        let id = self.new_node_id_for_identifier(declaration.identifier.clone());
        self.struct_declarations.insert(id, declaration);

        id
    }

    fn new_node_id_for_identifier<T: From<ID> + Into<NodeID> + Copy>(
        &mut self,
        name: Identifier,
    ) -> T {
        let id = self.new_id::<T>();
        self.ids.insert(id.into(), name.clone());
        id
    }

    fn new_id<T: From<ID> + Into<NodeID> + Copy>(&mut self) -> T {
        self.id_generator.new_id::<T>()
    }

    fn new_function_declaration(
        &mut self,
        declaration: FunctionDeclaration,
    ) -> FunctionDeclarationID {
        let id = self.new_node_id_for_identifier(declaration.identifier.clone());
        self.function_declarations.insert(id, declaration);

        id
    }

    fn new_block(&mut self, block: Block) -> BlockID {
        let id = self.new_id();
        self.blocks.insert(id, block);
        id
    }

    fn new_expression(&mut self, expression: Expression) -> ExpressionID {
        let id = self.new_id();
        self.expressions.insert(id, expression);
        id
    }
}
