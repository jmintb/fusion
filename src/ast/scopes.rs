use crate::ast::NodeID;
use crate::ast::ScopeID;
use std::collections::HashMap;

use super::identifiers::DeclarationID;
use super::identifiers::StatementID;
use super::Ast;
use super::NodeDatabase;

#[derive(Debug, Clone)]
pub struct Scope {
    pub nodes: Vec<NodeID>,
    pub parent_scope: Option<ScopeID>,
    pub associated_node: Option<NodeID>,
}

pub fn build_program_scopes(ast: &Ast, db: &NodeDatabase) -> HashMap<ScopeID, Scope> {
    let mut scopes: HashMap<ScopeID, Scope> = HashMap::new();

    let root_scope_id = db.new_id::<ScopeID>();
    let root_scope = Scope {
        nodes: Vec::new(),
        parent_scope: None,
        associated_node: None,
    };

    scopes.insert(root_scope_id, root_scope);
    let mut current_scope_id = root_scope_id;

    let mut que: Vec<NodeID> = ast
        .modules
        .clone()
        .into_iter()
        .map(|module_id| module_id.into())
        .collect();

    let node_relationships = ast.get_node_relationships(db);
    let mut scope_stack = vec![current_scope_id];

    while let Some(node_id) = que.pop() {
        scopes
            .get_mut(&current_scope_id)
            .unwrap()
            .nodes
            .push(node_id);

        match node_id {
            NodeID::Declaration(DeclarationID::FunctionDeclaration(_))
            | NodeID::Statement(StatementID::Declaration(DeclarationID::FunctionDeclaration(_)))
            | NodeID::Block(_) => {
                let new_scope = Scope {
                    nodes: Vec::new(),
                    parent_scope: Some(current_scope_id),
                    associated_node: Some(node_id),
                };

                let new_scope_id = db.new_id::<ScopeID>();
                scopes.insert(new_scope_id, new_scope);

                let current_scope = scopes.get(&current_scope_id).unwrap();
                while current_scope.associated_node.is_some()
                    && !node_relationships
                        .get(&current_scope.associated_node.unwrap())
                        .unwrap_or(&Vec::new())
                        .contains(&node_id)
                {
                    scope_stack.pop();
                }

                current_scope_id = new_scope_id;
                scope_stack.push(current_scope_id);

                for &child in node_relationships.get(&node_id).unwrap_or(&Vec::new()) {
                    que.push(child);
                }
            }
            _ => (),
        }
    }

    scopes
}
