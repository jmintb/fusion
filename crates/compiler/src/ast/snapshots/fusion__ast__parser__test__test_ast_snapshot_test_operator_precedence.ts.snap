---
source: src/ast/parser.rs
expression: "format!(\"{}\", ast.0.to_string(&ast.1))"
---
root, [
    [
        "module_24",
    ],
], [
    [
        "main",
    ],
], [
    [
        "block_22",
    ],
], [
    [
        "a = Int(2) ",
        "b = Int(4) ",
        "c = Int(3) ",
        "resultb = Operation(Operation(Variable(c) + Variable(a)) * Variable(b)) ",
        "outputb = \"resultb: %d \\n\" ",
        "call sprintf([\"Variable(outputb)\", \"Variable(outputb)\", \"Variable(resultb)\"])",
        "call print([\"Variable(outputb)\", \"Int(13)\"])",
        "return",
    ],
]
